### PR TITLE
app-layer-tls: SNI extension support

### PR DESCRIPTION
**Pull request by thus**
Decode TLS client hello handshake to get server name from SNI extension. Add this value to tls-log (extended), output-json-tls (extended), and make it available to lua scripts through TlsGetSNI().

Fixed changes suggested by @jasonish in #1684.

### CODE REVIEW

File: src/app-layer-ssl.c [hunk 1]
```diff
@@ -143,11 +146,97 @@ static int SSLv3ParseHandshakeType(SSLState *ssl_state, uint8_t *input,
...
             ssl_state->flags |= SSL_AL_FLAG_STATE_CLIENT_HELLO;
+
+            /* skip version */
+            input += SSLV3_CLIENT_HELLO_VERSION_LEN;
+
+            /* skip random */
+            input += SSLV3_CLIENT_HELLO_RANDOM_LEN;
+
+            /* skip session id */
+            uint8_t session_id_length = *(input++);
```
Comment (inliniac) [inline src/app-layer-ssl.c]: Here we use the input that we just moved forward, _before_ it is checked 2 lines later. Seems more sane to do it the other way around?

File: src/app-layer-ssl.c [hunk 2]
```diff
@@ -143,11 +146,97 @@ static int SSLv3ParseHandshakeType(SSLState *ssl_state, uint8_t *input,
...
+
+            if ((input - initial_input) > input_len)
+                goto end;
+
+            input += session_id_length;
+
+            /* skip cipher suites */
+            uint16_t cipher_suites_length = ntohs(*(uint16_t *)input);
```
Comment (inliniac) [inline src/app-layer-ssl.c]: same here

File: src/app-layer-ssl.c [hunk 3]
```diff
@@ -143,11 +146,97 @@ static int SSLv3ParseHandshakeType(SSLState *ssl_state, uint8_t *input,
...
+            input += 2;
+
+            if ((input - initial_input) > input_len)
+                goto end;
+
+            input += cipher_suites_length;
+
+            /* skip compression methods */
+            uint8_t compression_methods_length = *(input++);
```
Comment (inliniac) [inline src/app-layer-ssl.c]: and here

**Comment by inliniac** [comment]
Think the pointer logic is wrong. The pattern is now: move ptr fwd, use it, then check it. Should be move fwd, check it, then use it.

**Comment by thus** [comment]
When you mention it, I see what you mean. The logic (or the opposite of logic) in my head was thinking that I were dealing with pointers, and that I had to check them before I used them, but I'm dealing with integers here.

Coffee first, then new pull request! :)
(should always be applied in this order)

**Pull request closed without merging**